lib/mongodb.js

```diff
@@ -1996,23 +1997,74 @@ function optimizedFindOrCreate(model, filter, data, options, callback) {
 }
 
 /**
- * Convert the decimal properties from string to decimal.
- * Only supported after mongodb 3.4
  *
- * @param {Object} data The data that might contain a decimal property
- * @param {Object} props The model property definitions
+ * @param {*} data Plain Data Object for the matching property definition(s)
+ * @param {*} modelCtorOrDef Model constructor or definition
+ * @param {*} setValue A callback function which takes a property value and
+ * definition to apply custom property coercion
  */
-function convertDecimalProps(data, props) {
-  if (debug.enabled) debug('convertDecimalProps props: ', util.inspect(props));
-  for (const p in props) {
-    const prop = props[p];
-    const isDecimal = data[p] && prop && prop.mongodb &&
-      prop.mongodb.dataType &&
-      prop.mongodb.dataType.toLowerCase() === 'decimal128';
-    if (isDecimal) {
-      data[p] = Decimal128.fromString(data[p]);
-      debug('convertDecimalProps decimal value: ', data[p]);
+function visitAllProperties(data, modelCtorOrDef, setValue) {
+  if (data === null) return;
+  const modelProps = modelCtorOrDef.properties ? modelCtorOrDef.properties : modelCtorOrDef.definition.properties;
+  const allProps = new Set(Object.keys(data).concat(Object.keys(modelProps)));
+  for (const p of allProps) {
+    const value = data[p];
+    const def = modelProps[p];
+    if (def && def.type && isNestedModel(def.type)) {
+      if (Array.isArray(def.type)) {
+        if (value === null || value.length === 0) continue;
+        for (const it of value) {
+          visitAllProperties(it, def.type[0].definition, setValue);
+        }
+      } else {
+        if (value === null) continue;
+        visitAllProperties(value, def.type.definition, setValue);
+      }
+    } else {
+      data[p] = setValue(value, def);
```

Comment:
Continuing discussion started earlier. Instead of using the value returned by the visitor to update the property, we should provide the visitor function with a tool allowing the visitor to decide if the property value should be updated or not.

```suggestion
      visitor(value, def, (newValue) => { data[p] = newValue; });
```

Reply:
Agreed, this had unintended behaviour (backwards-incompatible) that was setting undefined values in the data object when we should skip them. Thank you!